File: tictactoe.rb
board = [[0, 0, 0], [0, 0, 0], [0, 0, 0]]
message = ""
current_player = 1  # 1 for X, 2 for O

def is_board_filled(board)
    board.flatten.none? { |cell| cell == 0 }
end

def print_board(board, message)
    puts "\x1b[H\x1b[2J"
    puts "\n  0 1 2"
    board.each_with_index do |row, i|
        print "#{i} "
        row.each do |cell|
            symbol = case cell
                        when 1 then 'X'
                        when 2 then 'O'
                        else '.'
                    end
            print "#{symbol} "
        end
        puts
    end
    puts message
end

def get_player_move(board, current_player, message)
    loop do
        print_board(board, message)
        print "Player #{current_player == 1 ? 'X' : 'O'}'s turn. Enter row and column (0-2): "
        input = gets.chomp.split.map(&:to_i)
        
        if input.length < 2
            message = "Invalid input. Enter two numbers."
            next
        end
        
        row, col = input[0], input[1]
        
        if row < 0 || row > 2 || col < 0 || col > 2
            message = "Position out of bounds. Try again."
        elsif board[row][col] != 0
            message = "Position occupied. Try again."
        else
            return [row, col], ""
        end
    end
end

def check_win(board, row, col)
    player = board[row][col]
    
    # Check row
    return true if board[row].all? { |cell| cell == player }
    
    # Check column
    return true if board.all? { |r| r[col] == player }
    
    # Check diagonals
    if row == col
        return true if (0..2).all? { |i| board[i][i] == player }
    end
    
    if row + col == 2
        return true if (0..2).all? { |i| board[i][2 - i] == player }
    end
    
    false
end

# Game loop
game_ended = false
until game_ended
    move, message = get_player_move(board, current_player, message)
    row, col = move
    board[row][col] = current_player
    
    if check_win(board, row, col)
        winner = current_player == 1 ? 'X' : 'O'
        message = "Player #{winner} wins!"
        print_board(board, message)
        game_ended = true
    elsif is_board_filled(board)
        message = "It's a draw!"
        print_board(board, message)
        game_ended = true
    else
        current_player = 3 - current_player  # Switch player
        message = ""
    end
end

File: tictactoe.lua
local board = {{0, 0, 0}, {0, 0, 0}, {0, 0, 0}}
local message = ""
local curr_player = 1  -- 1 for X, 2 for O

local function is_board_filled()
    for i = 1, 3 do
        for j = 1, 3 do
            if board[i][j] == 0 then return false end
        end
    end
    return true
end

local function print_board()
    print("\x1b[H\x1b[2J")
    print("\n  0 1 2")
    for i = 1, 3 do
        io.write(i - 1 .. " ")
        for j = 1, 3 do
            local cell = board[i][j]
            local symbol = "."
            if cell == 1 then
                symbol = "X"
            elseif cell == 2 then
                symbol = "O"
            end
            io.write(symbol .. " ")
        end
        print()
    end
    print(message)
end

local function get_player_move()
    while true do
        print_board()
        io.write(string.format("Player %s's turn. Enter row and column (0-2): ",
                curr_player == 1 and "X" or "O"))
        
        local input = io.read("*l")
        local row, col = input:match("(%d)%s(%d)")
        
        if not row or not col then
            message = "Invalid input. Enter two numbers."
        else
            row = tonumber(row) + 1  -- Lua arrays are 1-based
            col = tonumber(col) + 1
            
            if row < 1 or row > 3 or col < 1 or col > 3 then
                message = "Position out of bounds. Try again."
            elseif board[row][col] ~= 0 then
                message = "Position occupied. Try again."
            else
                message = ""
                return row, col
            end
        end
    end
end

local function check_win(row, col)
    local player = board[row][col]
    
    -- Check row
    if board[row][1] == player and board[row][2] == player and board[row][3] == player then
        return true
    end
    
    -- Check column
    if board[1][col] == player and board[2][col] == player and board[3][col] == player then
        return true
    end
    
    -- Check diagonals
    if row == col then
        if board[1][1] == player and board[2][2] == player and board[3][3] == player then
        return true
        end
    end
    
    if row + col == 4 then  -- Because 1-based: (1,3)=4, (2,2)=4, (3,1)=4
        if board[1][3] == player and board[2][2] == player and board[3][1] == player then
        return true
        end
    end
    
    return false
end

-- Game loop
local game_ended = false
while not game_ended do
    local row, col = get_player_move()
    board[row][col] = curr_player
    
    if check_win(row, col) then
        local winner = curr_player == 1 and "X" or "O"
        message = string.format("Player '%s' wins!", winner)
        print_board()
        game_ended = true
    elseif is_board_filled() then
        message = "It's a draw!"
        print_board()
        game_ended = true
    else
        curr_player = 3 - curr_player  -- Switch player
    end
end